File: leetcode/edit_distance/EditDistance.cpp
#include <iostream> 

using namespace std; 

class Solution {
public:
    int min(int a, int b, int c) {
        int min = a; 
        if (b < min) {
            min = b; 
        }
        
        if (c < min) {
            min = c; 
        }
        
        return min; 
    }
    
    int minDistance(string word1, string word2) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        
        int T[word1.size() + 1][word2.size() + 1];  // T[i][j]: edit distance of first ith word1 and first jth word2
        
        T[0][0] = 0; 
        
        // Initialize first row
        for (int i = 1; i <= word2.size(); i++) {
            T[0][i] = T[0][i - 1] + 1; 
        }
        
        // Initialize first column
        for(int i = 1; i <= word1.size(); i++) {
            T[i][0] = T[i - 1][0] + 1; 
        }
        
        for (int i = 1; i <= word1.size(); i++) {
            for (int j = 1; j <= word2.size(); j++) {
                if (word1[i - 1] == word2[j - 1]) {
                    T[i][j] = min(T[i - 1][j - 1], T[i - 1][j] + 1, T[i][j - 1] + 1);  
                }
                else {
                    T[i][j] = min(T[i - 1][j - 1] + 1, T[i - 1][j] + 1, T[i][j - 1] + 1); 
                }
            }
        }
        
        return T[word1.size()][word2.size()];
    }
};

// Test Driver 
int main() {
    Solution sln; 
    cout << sln.minDistance("sea", "eat") << endl; 
    return 0; 
}

File: leetcode4/EditDistance.java
public class Solution {
    public int minDistance(String word1, String word2) {
        int m = word1.length(); 
        int n = word2.length(); 
        
        int[][] T = new int[m + 1][n + 1];
        T[0][0] = 0; 
        
        for (int i = 1; i < m + 1; i++) {
            T[i][0] = i; 
        }
        
        for (int j = 1; j < n + 1; j++) {
            T[0][j] = j; 
        }
        
        for (int i = 1; i < m + 1; i++) {
            for (int j = 1; j < n + 1; j++) {
                if (word1.charAt(i - 1) == word2.charAt(j - 1)) {
                    T[i][j] = T[i - 1][j - 1]; 
                } else {
                    T[i][j] = T[i][j - 1] + 1; // add 
                    T[i][j] = Math.min(T[i][j], T[i - 1][j] + 1);  // delete
                    T[i][j] = Math.min(T[i][j], T[i - 1][j - 1] + 1);  // update 
                }
            }
        }
        
        return T[m][n]; 
    }
}
